### ISSUE TITLE
legacy-stream cuMemcpy races non-blocking streams: 260 host-transfer sites assume an ordering CUDA does not provide

### ISSUE THREAD

**Issue by noahgift** [paiml/aprender#2767]
Root cause of #2753's residue, and it is systemic rather than a slip.

`CudaStream::new` creates this crate's streams with **`CU_STREAM_NON_BLOCKING`**, which CUDA explicitly excludes from legacy default-stream ordering. `GpuBuffer::copy_from_host` and `copy_to_host` are **`cuMemcpyHtoD` / `cuMemcpyDtoH` — legacy-stream transfers**.

A legacy transfer does not order against a non-blocking stream. So every host transfer in the batched decode path races the kernels in flight, while the surrounding code is written throughout as though the transfers were ordered against it.

## Proven by mutation, one binary, one env var

```
CU_STREAM_NON_BLOCKING (default)          SELF cosine 0.517985   FAILS
APR_STREAM_LEGACY=1 (CU_STREAM_DEFAULT)   SELF cosine 1.000000   PASSES
```

SELF runs the batched forward twice on the same input and the same state and requires identical logits. This **restores the invariant** rather than improving a number. SENSITIVE and PERTURBED still diverge under the knob, so the probe has not gone uniformly green.

One mechanism accounts for the entire #2753 thread: the allocation-size dependence (timing), the all-NaN and all-zero logits (a partially written buffer; a fresh allocation reads as zero, which argmaxes to token 0 — the original constant-token garbage), and the first-call/second-call asymmetry.

## The scale

**260 legacy-stream host-transfer sites in `aprender-serve` alone** — 206 `copy_to_host`, 31 `copy_from_host`, 15 `copy_from_host_at`, 8 `copy_to_host_at` — including the pointer arrays, `seq_lens` and `positions` that in-flight kernels are reading.

## Two repairs, and why the default was left alone

**(a) Create the streams with `CU_STREAM_DEFAULT`.** One line. Restores the ordering the code already assumes — and **serialises the overlap**. This crate creates several streams (`compute_stream`, `transfer_stream`, `s1..s3`) deliberately for concurrency, and CORRECTNESS-012b moved work onto `self.stream` precisely for ordering. Flipping them all to the legacy stream serialises them against it.

**(b) Convert the transfers to the stream-ordered async variants that already exist** in `transfer.rs` (`copy_from_host_async`, `cuMemcpyDtoHAsync`), with an explicit sync before each host read. Preserves the overlap. 260 call sites.

The default was deliberately left unchanged rather than flipped, because (a) has a throughput consequence that nobody has measured and this is a performance-critical path.

## Recommendation

**Take (a) first as the correctness fix, measure the throughput cost, and let that number decide whether (b) is warranted.**

A wrong-but-fast path is not a tradeoff, it is a defect — and the measured cost of (a) is exactly the evidence needed to justify (b)'s 260-site conversion, or to show it is unnecessary. Doing (b) first means a large mechanical change on the assumption that (a) is too slow, which is the kind of unmeasured premise APR-PERF-GATE-001 (#2706) exists to remove.

Note the interaction with #2764: throughput measured at `m >= 3` before that fix also included a per-layer synchronize and D2H copy, so any historical baseline used to argue about (a)'s cost needs re-taking anyway.

**Comment by noahgift** [comment]
## Measured: repair (a) costs 0% at c=1 and 11–18% at c=4–8, rising with concurrency

Not negligible — so this does **not** retire (b). But the correctness side turned out to be much
worse than "a wrong-but-fast path", which changes the ordering rather than the conclusion.

### The number

`agg_tok_s` (§4.4.3 wall-clock aggregate), median of **3 replicates** per cell, one binary,
one knob:

| band | default (`CU_STREAM_NON_BLOCKING`) | `APR_STREAM_LEGACY=1` (repair **a**) | ratio | verdict |
|---|---|---|---|---|
| c=1 | 328.87 | 330.99 | **1.006** | replicate ranges **overlap** → no measurable cost |
| c=4 | 398.75 | 356.07 | **0.893** (−10.7%) | ranges **disjoint** |
| c=8 | 1484.05 | 1211.90 | **0.817** (−18.3%) | ranges **disjoint** |

"Disjoint" = every legacy replicate fell below every default replicate. Per-replicate values:

```
c=1  nb 328.83 332.11 328.87   |  legacy 330.99 342.23 320.93
c=4  nb 389.01 404.88 398.75   |  legacy 356.07 357.18 347.66
c=8  nb 1484.05 1533.87 1452.85|  legacy 1241.92 1211.90 1146.11
```

The cost is **zero at c=1 and monotone in c**. c=16 was not measured; the trend says it is worse
than 18%.

### The correctness side, which reframes the decision

Same binary, same knob, at the level a user sees. 10 rounds x 4 concurrent **identical** greedy
requests (`temperature=0`, `max_tokens=64`), 40 responses per arm. A correct greedy decoder must
return one distinct string.

| arm | responses | distinct continuations | agree with modal |
|---|---|---|---|
| default | 40 | **11** | 10.0% |
| `APR_STREAM_LEGACY=1` | 40 | **1** | 100.0% |

Most of the default's outputs are not merely different, they are garbage:

```
'0  IR;;";;;; N N1, NATL()000 first,0_0/"、、 first first'
'*/9,9,0'
'C`.`A`.`'
'»DCI(I_N01 ===,,,,,,,,100// _st_1.//,,,,,,,,(//0  ===II_0_1-level('
```

...against the single legacy answer, which is the correct one
(`"Compilers are software programs that translate high-level programming languages into machine
code..."`). Within a round all 4 slots agree; across rounds the answer changes — the corruption is
**per-batch, not per-slot**.

This doubles as the engagement proof for the knob in *this* binary: a behavioural delta, not a
label on a run.

### Protocol (stated plainly — this is NOT a §4.4 receipt)

**Conformant per band**: closed-loop, external HTTP, fixed c, 2/worker warmup discarded, 5 s
quiesce, terminate on the later of `max(30, 8c)` samples and 60 s, wall-clock aggregate,
server-reported `usage.completion_tokens` (`tokenization.method = server_usage`). Client is the
repo's own `scripts/perf041_client.py`, unmodified. Generation follows W1 §4.3.1:
`max_tokens=128`, greedy, context 4096.

**Not conformant**: N=3 replicates but **no §4.4.4 bootstrap CI** — the client retains no
per-request samples, so I report every replicate and its range instead of a resampled interval;
a single fixed prompt rather than the W1 512-token corpus; the **1.5B** model, not W1's 7B (see
below); no receipt emitted; `drain_ms` not measured. Calling this a §4.4 receipt would be false.
It is a within-host A/B of one knob, which is what the question needs.

`scripts/perf-matrix.yaml` untouched; all 8 cells remain UNMEASURED.

### Controls

- **One binary, both arms.** The knob is the only difference.
- **Arms interleaved** (nb, legacy, nb, legacy, ...) within each replicate, so host drift is not
  confounded with the arm. Load during the run: median 5.25, max 32.17 (shared box).
- **#2764**: `APR_LAYER_TRACE` unset; verified **0** `[CB-006-OUT]` lines across all 6 server logs.
- **#2770**: `CUBLAS_GEMM_THRESHOLD=32` on **both** arms, so neither run is on the cuBLAS decode
  route that diverges at m>=4.
- **#2768 — binary pinned by content, not `--version`**: sha256
  `33f2c50c78b40820815c1effbefe126ec76fe1d16d5a002cea543ea439c17d30`.
  `strings -a` finds `APR_GRAPH_NO_BIAS` in it (proves bf0ca8249 / #2769 is compiled in) and
  **absent** from the shared-target-dir build a plain `cargo build` would have produced — the
  shell's `cargo` function exports a shared `CARGO_TARGET_DIR`, so an explicit `--target-dir` was
  used. Built from `f27b3571c` (feat/v9-batchfix tip), `--features cuda`.
- Host lambda, RTX 4090 sm_89, CUDA 12.8. Model `qwen2.5-coder-1.5b-instruct-q4_k_m.gguf`
  sha256 `cc324af070c2ecbfd324a30884d2f951a7ff756aba85cb811a6ec436933bb046`.

### Why there is no W1 (7B) number at c>1

An independent blocker, on **both** arms:

```
[PMAT-072] Setup+prefill ERROR (m=3): Operation 'init_batched_kv_cache_gpu' not supported:
Failed to init batched KV cache for M=32: GPU memory allocation failed:
CUDA driver error: CUDA_ERROR_OUT_OF_MEMORY (code: 2)
```

It preallocates for **M=32 regardless of the actual m** (seen at m=3 and m=8), which does not fit
a 24 GB 4090 at 7B/ctx4096. c=1 works. So apr cannot serve W1 batched on lambda at all, on either
arm — unrelated to streams, and probably worth its own issue.

### What the cost is *not*

The issue attributes (a)'s cost to serialising "the overlap this crate creates deliberately". I
could not find that overlap in production. PARITY-038's machinery — `copy_to_gpu_async`,
`copy_from_gpu_async`, `synchronize_transfer`, `synchronize_all` — and `CudaPipeline` have **zero
non-test callers**, so `transfer_stream` carries no production work at all. (Also:
`compute_stream`, `transfer_stream` and `stream` are three *distinct* streams — the
"Legacy alias for compute_stream" comment on `stream` in `executor/mod.rs` is wrong.)

The cost being zero at c=1 and growing with c points instead at **cross-request** concurrency:
`checkout_streams` mints a fresh triple whenever the single-entry `STREAM_POOL` is empty, so
concurrent executors get independent non-blocking streams that `CU_STREAM_DEFAULT` then
serialises against one another. **I did not prove this mechanism** — it is the explanation most
consistent with the c-dependence, not a measured cause.

### Recommendation: take (a) now, and (b) is warranted — but try one cheaper thing first

1. **Ship (a) as the default immediately.** The default returns 11 different answers to the same
   greedy question and most of them are garbage. 18% is not a tradeoff against that; it is the
   price of returning the right answer at all.
2. **(b) is justified by the number, not assumed.** Up to 18% aggregate at c=8 and rising, on
   exactly the bands Arm A and B1 gate. That is the evidence #2767 asked for.
3. **Correct the site count**: 260 includes test files. Production-only is **195**
   (146 `copy_to_host`, 26 `copy_from_host`, 15 `copy_from_host_at`, 8 `copy_to_host_at`).
4. **Falsify the cheap fix before doing 195 sites.** If the cross-request hypothesis above holds,
   pooling streams per-executor (or sharing one stream across concurrent executors) recovers most
   of the 18% for a handful of lines. That is a one-day experiment against a 195-site conversion,
   and it is testable with the same A/B harness used here.

Also visible in the same data, separate from this issue: **c=4 is anomalously bad on both arms.**
Scaling efficiency is 0.30 at c=4 but 0.56 at c=8, and p50 latency is *higher* at c=4 (1.24 s)
than at c=8 (0.67 s). Efficiency should not rise from c=4 to c=8.

---

Measurement only — no code changes, so no branch was pushed. The `APR_STREAM_LEGACY` knob already
on `feat/v9-batchfix` was sufficient to run both arms from one binary.

**Comment by noahgift** [comment]
## Decision: ship (a) now. The number justifies (b) rather than assuming it.

The measurement is posted above. Recording the decision it supports, with its dissent, so this is not re-litigated.

**Repair (a) is not free** — −10.7% at c=4, −18.3% at c=8, every legacy replicate below every default replicate, ranges disjoint. My earlier hope that it might be negligible is refuted.

**Ship it anyway, today.** Ten rounds of four concurrent *identical greedy* requests: the default returns **11 distinct continuations out of 40**, most of them garbage; repair (a) returns **1 out of 40**, the correct answer. That is not a performance/correctness tradeoff to be weighed. A server that answers the same deterministic question 40 times and gives 11 different answers is broken, and 18% of a wrong number is worth nothing.

**Then do (b) to recover the throughput.** This is exactly the sequencing I asked for and did not assume: the conversion is now justified by a measured 18%, rather than by a guess that (a) would be too slow. Before starting it, note two corrections the measurement made to this issue's own premises:

- **The overlap (a) supposedly serialises does not exist in production.** `copy_to_gpu_async`, `copy_from_gpu_async`, `synchronize_transfer`, `synchronize_all` and `CudaPipeline` have **zero non-test callers**, and `transfer_stream` carries no production work. The cost is zero at c=1 and grows with c, which points instead at cross-request concurrency (`checkout_streams` minting fresh triples). That explanation is **unproven** — it is the one most consistent with the c-dependence, not a measured cause — but it is cheap to falsify and **should be falsified before touching 195 sites**, because if it is right the fix is far smaller than (b).
- **The 260 figure in this issue counts test files.** Production-only is **195**.

**Protocol, stated rather than implied:** §4.4-conformant per band (closed-loop external HTTP, 2/worker warmup, 5 s quiesce, `max(30,8c)` samples and ≥60 s, server-reported token counts, via the repo's own `perf041_client.py`). **Not** a §4.4 receipt — no §4.4.4 bootstrap CI, single prompt rather than the W1 corpus, 1.5B rather than W1's 7B. Every replicate and range is reported instead.

Controls that make it a comparison rather than two numbers: `APR_LAYER_TRACE` off with 0 `[CB-006-OUT]` lines verified in all six server logs (#2764), `CUBLAS_GEMM_THRESHOLD=32` on **both** arms (#2770), arms interleaved against host drift, and the binary pinned **by content** — which caught #2768's trap live, since the shared-target build lacked `APR_GRAPH_NO_BIAS` and would have been the wrong binary.

The W1 gap has its own cause and its own ticket: **#2774**, the batched KV cache preallocating for M=32 regardless of actual batch, which OOMs the 7B model on a 24 GB 4090.

**Comment by noahgift** [comment]
Closing — fixed on main: fixed by PR #2776 (merged 2026-08-31): streams are created CU_STREAM_DEFAULT by default (aprender-gpu driver/memory/transfer.rs:48-55; APR_STREAM_NONBLOCKING=1 restores the racy flag as a measurement knob); the cost table (0.893×/0.817× at c=4/8) and the 11→1 distinct-continuation result are in that PR's body and the transfer.rs doc (l.120-131). The (b) 195-site async conversion to recover throughput is 0.67 speed work (§6 W-E scheduler), stated in the close comment

Repair (a) — the decision the issue's own comment recorded — is on main; the measured cost is the receipt.

Reopen if the pointer does not answer it.

_PP-066 triage 2026-09-05 (issue #2873); ledger `docs/audits/pp-066-triage-2026-09-05.md`._
